unlocked update & draw rate, delta time, new example to demonstrate

Byter09 opened the pull request:
<!--- Describe your changes in detail -->
I'm using quicksilver for physics based stuff and rely on a very high update rate. This change unlocks both the update and the draw rate, which additionally are fully seperated now.
## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
With the unlocked draw and update rate, one can implement application specific timings.
Additionally, if required, the default values of quicksilver would allow 60 fps and 60 updates per second.

Oh and I upgraded `current_time()` from using just milliseconds to nanoseconds on desktop, because time difference calculation was way to imprecise without it.

Possible unwanted changes: I removed the max_ticks, because I thought they're not required anymore. I could not think of a use case for them. If, however, one would need to execute `n` update calls and then one draw call, this change removes that feature from quicksilver and allows the user to implement it themself. If it is required to have this feature back, I'll add it back with an additional commit to this branch/PR.
<!--- If it fixes an open issue, please link to the issue here. -->
This fixes issue #337.

## Screenshot:
<!--- You can drag image files into GitHub's edit-window -->
![stopwatch](https://user-images.githubusercontent.com/16060697/44366938-954efc80-a4ce-11e8-92f9-b9281aae90a4.png)

## Types of changes
<!--- What types of changes does your code introduce? Remove all those that do not apply -->
- New feature (non-breaking change which adds functionality)
- Bug fix (non-breaking change which fixes an issue)

## Checks
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read `CONTRIBUTING.md`.
- [x] This Pull Request targets the right branch 
- [x] I have updated `CHANGES.md`, with [BREAKING] next to all breaking changes
- [x] I have updated the documentation accordingly if necessary
- [x] I have updated / added tests to cover my changes if necessary
<!-- Remove these checks if this Pull Request does not affect the public API -->
- [x] The example found in `README.md` compiles and functions like expected
- [x] The example found in `src/lib.rs` compiles and functions like expected

ryanisaacg commented:
This looks pretty good overall! I have a few nitpicks and changes, but I'm not currently able to give a thorough code review; I should be able to tonight.

Byter09 commented on CHANGES.md:
```diff
@@ -35,11 +35,12 @@
 - [Breaking] Move `State` and `run` into a new `lifecycle` module
 - Add functions to determine the current framerate
-- Add the ability to customize the tick rate
+- Add the ability to customize the update and tick rate independently
```
I choose "rate" because this PR would change the fixed timestep to a more open but controllable method, meaning there are no "ticks" anymore. It just happens as fast as possible and with a defined delay. If your calculation is fast enough between update/draw calls nothing will happen until this delay is reached, or if it is not fast enough and for example your update call takes longer than expected it'll simply catch up again as the delay is reached anyway.

Byter09 commented on examples/stopwatch.rs:
```diff
@@ -0,0 +1,94 @@
+        let minutes = ((self.elapsed / 1000.) / 60.) % 60.;
+        let hours = ((self.elapsed / 1000.) / 60. / 24.) % 24.;
+
+        self.seconds = seconds;
```
That is correct and something I missed. There's probably a bit more but this is why I said I'll push additional stuff.

Byter09 commented on examples/stopwatch.rs:
```diff
@@ -0,0 +1,94 @@
+        let minute_angle = 2. * PI * ((self.minutes+45.) / 60.);
+        let second_angle = 2. * PI * ((self.seconds+45.) / 60.);
+
+        let hour_pos_x = 400. + 150. * hour_angle.cos() as f32;
```
Indeed. I oversaw that. I should've looked at the API beforehand.

Byter09 commented on examples/stopwatch.rs:
```diff
@@ -0,0 +1,94 @@
+
+
+
+        let hour = Line::new((400, 300), (hour_pos_x, hour_pos_y)).with_thickness(10.);
```
`with_thickness()` requires a float. So why would `10.` be incorrect?

ryanisaacg replied:
Oh, I thought it was `impl Scalar`. It probably should be for consistency, but for now `10.` is perfectly correct.

Byter09 replied:
Yeah that is still my fault, I wrote `with_thickness()` but I can not remember why I used f32 there... Shall I also change that?

Byter09 commented on examples/stopwatch.rs:
```diff
@@ -0,0 +1,94 @@
+}
+
+fn main() {
+    let mut settings = Settings::default();
```
Oh, because my way allocated a value twice. Yes. Indeed.

Byter09 commented on src/lifecycle/application.rs:
```diff
@@ -37,26 +35,27 @@ impl<T: State> Application<T> {
-        self.window.log_framerate(current - self.last_draw);
-        self.last_draw = current;
+        let delta_draw = current - self.last_draw;
+        if delta_draw >= self.window.draw_rate() {
```
Indeed there is. Just a small example would be a mini-wasm clock or something else that'll only update every second or so on your website. Why should it draw more often?

Or imagine FPS limiting? Even if it uses vsync (which I totally forgot), why would I not want my application to limit to 30fps if more is not necessary?

ryanisaacg replied:
Both of these sound like good use cases, 👍 to variable draw rate

Byter09 commented on src/lifecycle/state.rs:
```diff
@@ -13,21 +13,21 @@ pub trait State: 'static {
     ///
     /// By default it does nothing
-    fn update(&mut self, &mut Window) -> Result<()> {
+    fn update(&mut self, window: &mut Window, delta_time: f64) -> Result<()> {
```
Indeed it could be. But having to assign a delta time to the window each time instead of just giving it through a parameter seemed more efficient and clean. Also, because there are multiple delta_times (I could even imagine one for events but the user could easily implement that themself, as it has no value to quicksilver), we would have multiple operations more where only a simple pass is required.

ryanisaacg replied:
True, there would be some definite weirdness unless it's a parameter.

Byter09 commented on src/lifecycle/state.rs:
```diff
@@ -13,21 +13,21 @@ pub trait State: 'static {
     ///
     /// By default it does nothing
-    fn event(&mut self, &Event, &mut Window) -> Result<()> {
+    fn event(&mut self, event: &Event, window: &mut Window) -> Result<()> {
```
Anonymous method parameters are obsolete as of Rust 2018 (and a bit before that I believe).

ryanisaacg replied:
Oh, thanks for the heads up!

Byter09 commented on src/lifecycle/application.rs:
```diff
@@ -12,7 +12,6 @@ pub struct Application<T: State> {
     pub state: T,
     pub window: Window,
     pub event_buffer: Vec<Event>,
-    accumulator: f64,
```
This was indeed something I was thinking about but the sheer fact, that the `update()` method was called inside a while loop was just so bizarre.
In my use case for example, it cause the drawing to lag behind, even though there was enough time to draw. I think my settings were: `max_ticks = 10000`, `tick_rate = 1/60.`

Even though the value is called "max_ticks" quicksilver ALWAYS called `update()` 10000 times instead of aborting when it was time to draw again. This caused massive lags and made my application useless in combination with quicksilver. Thus this PR.

Byter09 commented:
Just fixed the example to use `Vector::from_angle`, (and I just noticed, that I forgot something) and fixed the `with_thickness` method to use `impl Scalar`.

Byter09 commented:
Alright, a friend of mine took the time to explain to me the current accumulator implementation in detail and also re-implemented it together with all my other changes. The only thing, that's now a problem is the 100% CPU usage when we're looping over `draw` in case it is not time to draw again (because of a lowered draw_rate for example). Kind of feeling dumb for not understanding the accumulator at first but I get it now.

We (or he if I'm too dumb) will look into fixing that. If thats done, do you think this is a good PR? I mean basically we went back to the previous implementation I guess... but better? :P

ryanisaacg commented on src/lifecycle/application.rs:
```diff
@@ -37,26 +37,53 @@ impl<T: State> Application<T> {
-            self.state.update(&mut self.window)?;
+        let update_rate = self.window.update_rate();
+        while self.accumulator > 0.0 && (self.window.max_updates() == 0 || ticks < self.window.max_updates()) {
+            self.state.update(&mut self.window, update_rate)?;
```
Is there still a reason to pass in dt if it's just `window.update_rate()`?

Byter09 replied:
Probably not. I guess that can go then?

Byter09 replied:
Oh nevermind, I guess using
- [Breaking] Remove `WindowBuilder` and create a new `Settings` struct to replace it
is enough. Or do you want vsync to have a separate entry?

ryanisaacg replied:
@Byter09 I think vsync probably warrants its own entry in the changelog
@999eagle I would say probably not pass in DT, because as of right not it would just be duplicating `window.update_rate()` at present.

ryanisaacg commented on examples/stopwatch.rs:
```diff
@@ -0,0 +1,81 @@
+// Draw some multi-colored geometry to the screen
```
Forgot to update a comment? :P

Byter09 replied:
Oh man. I forgot so much stuff. I'll push a commit changing that and adding the changelog entry.

ryanisaacg commented:
Looks good to me! Is this ready to merge?

Byter09 commented:
Yeah. Nothing left to do I believe. Hope this PR is at least a bit helpful :D

ryanisaacg commented:
Very much so! I appreciate it 👍
